File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/1071-greatest-common-divisor-of-strings-approach2.py
from math import gcd

def gcd_of_strings(str1: str, str2: str) -> str:
    if str1 + str2 != str2 + str1:               # O(m+n) concatenate and compare
        return ""
    return str1[:gcd(len(str1), len(str2))]

def _run_tests():
    assert gcd_of_strings("ABCABC", "ABC") == "ABC"
    assert gcd_of_strings("ABABAB", "ABAB") == "AB"
    assert gcd_of_strings("LEET", "CODE") == ""
    assert gcd_of_strings("AAAAAB", "AAA") == ""
    assert gcd_of_strings("A", "A") == "A"
    assert gcd_of_strings("AAAA", "AA") == "AA"
    import time as _t
    _t0 = _t.perf_counter()
    gcd_of_strings("A" * 1000, "A" * 999)
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf gcd_of_strings n=1999: {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/1071-greatest-common-divisor-of-strings-approach2.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func gcdOfStrings(str1 string, str2 string) string {
	if str1+str2 != str2+str1 {
		return ""
	}
	return str1[:gcd(len(str1), len(str2))]
}

func runTests() {
	assert(gcdOfStrings("ABCABC", "ABC") == "ABC", "test 1")
	assert(gcdOfStrings("ABABAB", "ABAB") == "AB", "test 2")
	assert(gcdOfStrings("LEET", "CODE") == "", "test 3")
	assert(gcdOfStrings("AAAAAB", "AAA") == "", "test 4")
	assert(gcdOfStrings("A", "A") == "A", "test 5")
	assert(gcdOfStrings("AAAA", "AA") == "AA", "test 6")
	fmt.Println("all tests pass")
}

func main() { runTests() }
